vim.ui.open: "gx" without netrw

mrshmllow opened the pull request:
Fix #23231

`vim.ui.os_open` is partially based off https://github.com/folke/lazy.nvim/blob/main/lua/lazy/util.lua#L13-L44, as recommended by @folke on the original issue.

- [x] Implement `vim.ui.os_open(uri: string | vim.uri.Uri)`
    - [x] Accept a `string` as `uri`
    - [ ] ~~Accept a `vim.uri.Uri` as `uri`~~
- [x] Implement "gx" in lua. The implementation drops current netrw options, including `g:netrw_gx`.
- [ ] Ability to change impl by using `vim.func.on_func(vim.ui, 'open_uri', …)`

mrshmllow committed:
feat(vim.ui): vim.ui.open, "gx" without netrw

Co-authored-by: Mathias Fußenegger <mfussenegger@users.noreply.github.com>
Co-authored-by: Justin M. Keyes <justinkz@gmail.com>
Co-authored-by: ii14 <59243201+ii14@users.noreply.github.com>

glepnir commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,40 @@ function M.input(opts, on_confirm)
+    cmd = { 'xdg-open', uri }
+  elseif vim.fn.executable('wslview') == 1 then
+    cmd = { 'wslview', uri }
+  else
```
there should use system to check . not check executable .  like 
```lua
local ismac  = vim.loop.os_uname().sysname == 'Darwin'
-- windows check also like ismac.

if ismac then
     cmd = 'open'
end
```

glepnir commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,40 @@ function M.input(opts, on_confirm)
+---
+--- Example:
+--- <pre>lua
+--- vim.ui.input("https://neovim.io/")
```
hmm you forget change it . `vim.ui.open_uri` I am not sure it should in `ui` :)

glepnir commented:
the `uri` need check include some symbols or not ,  like `#` .  in some lsp hover response the link like `file:///usr/local/share/doc/xxx/System-IO.html#v:putStrLn` .  so you need use `vim.fn.escape` to escape `#` in uri .

clason commented:
> Is deleting code from runtime/plugin/netrwPlugin.vim appropriate / allowed?

No, please leave it as is (for now). It's better to set the corresponding global variables to skip setting up the mapping.

We'll completely remove netrw once we have a full replacement.

> What is the correct way to add new built-in keymaps in lua? The best location I found was runtime/plugin/nvim.lua.

Yes, that would be the place. But maybe make it opt-in for now?

mrshmllow commented:
God, the commit history in this PR is a complete mess lmao

glepnir commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,47 @@ function M.input(opts, on_confirm)
+  local cmd
+
+  local ismac = vim.loop.os_uname().sysname == 'Darwin'
+  local iswin = vim.loop.os_uname().sysname:lower():find('windows')
```
I wonder there just use `sysname:match('Windows')` not work for all windows version ?

mrshmllow replied:
this is how it is done in other parts of the code base

glepnir replied:
and you can just run `vim.loop.os_uname().sysname` this once right 
 ```lua
  local sysname = vim.loop.os_uname().sysname
  if sysname == 'Darwin' then
        open 
  elseif sysname:match('Windows') then
        foo
  else 
       bar 
  end
  ```

mrshmllow commented:
> No, please leave it as is (for now). It's better to set the corresponding global variables to skip setting up the mapping.

I am confused about the execution order, where would you set a global variable in time? I cannot find a place where this is done

> the uri need check include some symbols or not , like # . in some lsp hover response the link like file:///usr/local/share/doc/xxx/System-IO.html#v:putStrLn . so you need use vim.fn.escape to escape # in uri .

Im not entirely sure what you mean. Most applications will not understand what `#v:putStrLn` means. You can try to maybe adjust `g:netrw_gx` to capture the full #? (as `<cfile>` rightfully does not expect a file to have #)

glepnir commented:
it's a markdown link return by some lsp server. it has a `#` target symbol . but maybe it should not escape in `open_uri` function.  if I want use `open_uri` to open it i need escape `#` symbol by myself :)

clason commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,47 @@ function M.input(opts, on_confirm)
   end
 end
 
+--- Opens a URI by sending it to the system's default handler
```
```suggestion
--- Opens a URI by sending it to the system's default handler.
```

clason commented:
> I am confused about the execution order, where would you set a global variable in time? I cannot find a place where this is done

In your `init.vim`.  The point is only to add the functionality for now so users can use that instead of `netRW` _if they choose_. In a second step (after the early access crowd has found the majority of bugs), we can make this the default.

mfussenegger commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,45 @@ function M.input(opts, on_confirm)
+---
+---
+---@param uri string Uri to be opened
+---   TODO: uri should accept vim.uri.Uri
```
```suggestion
```

There is currently no `vim.uri.Uri` class. I think this can be removed for now.

mrshmllow replied:
@justinmk is accepting a Uri class really that important? There are no actions performed on it and all the handlers deal with parsing the "uri".

mfussenegger commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,45 @@ function M.input(opts, on_confirm)
+    cmd = 'xdg-open'
+  end
+
+  local ret = vim.fn.jobstart({
```
Could this return the `ret` and maybe have a second `opts` parameter with `quiet` to surpress the message? Then it could be used in https://github.com/neovim/neovim/blob/02d5a678fbf4ed36b4896010a051f136d587e779/runtime/lua/vim/lsp/handlers.lua#L571-L580

mrshmllow replied:
It might also be good to have a confirmation for security.

mrshmllow replied:
> I'm not sure about the opts/quiet, maybe it's fine to always have the error notification.

`system()` and `jobstart()` will complain either way.

mfussenegger commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,66 @@ function M.input(opts, on_confirm)
+    cmd = 'xdg-open'
+  end
+
+  if opts.confirm then
+    local answer =
+      vim.fn.confirm('Open "' .. uri .. '" using ' .. cmd .. '?', '&Yes\n&No', 2, 'Question')
+
+    if answer == 0 or answer == 2 then
+      return
+    end
+  end
```
I'd leave this out of this PR as it is a behavior change for LSP and it's not clear to me that having this confirm dialog is an improvement.

And it also seems to me that there's no good reason to build  this check into `open_uri` instead of doing it outside.

mfussenegger commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,66 @@ function M.input(opts, on_confirm)
+  if sysname == 'Darwin' then
+    cmd = 'open'
+  elseif sysname:match('windows') then
+    cmd = 'wslview'
```
Is `wslview` always available on windows?

mrshmllow replied:
Is there a good way to detect wsl or default windows cmd? Otherwise on the windows host, `start` also works I think.

mrshmllow replied:
Apparently [`/proc/sys/fs/binfmt_misc/WSLInterop`](https://superuser.com/questions/1749781/how-can-i-check-if-the-environment-is-wsl-from-a-shell-script#:~:text=The%20presence%20of,WSL%20at%20startup.) is a good indicator

justinmk commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,49 @@ function M.input(opts, on_confirm)
+function M.open_uri(uri)
+  local cmd
+
+  if vim.fn.has("wsl") then
```
Isn't `== 1` required here (and below)?

```suggestion
  if vim.fn.has("wsl") == 1 then
```

justinmk commented on runtime/plugin/nvim.lua:
```diff
@@ -18,3 +18,11 @@ vim.api.nvim_create_user_command('InspectTree', function(cmd)
   end
 end, { desc = 'Inspect treesitter language tree for buffer', count = true })
+
+if vim.g.use_lua_gx ~= nil then
```
@clason I think we can make this opt-out rather than opt-in from the start, it's low risk.

justinmk commented on runtime/plugin/nvim.lua:
```diff
@@ -18,3 +18,11 @@ vim.api.nvim_create_user_command('InspectTree', function(cmd)
+
+if vim.g.use_lua_gx ~= nil then
+  vim.keymap.set({ 'n', 'x' }, 'gx', function()
+    local uri = vim.fn.expand(vim.g.netrw_gx ~= nil and vim.g.netrw_gx or '<cfile>')
```
The purpose is to avoid netrw so supporting `vim.g.netrw_gx` is out of scope. For now we can just use `<cfile>`, eventually we might want to introduce `<curi>`.

justinmk commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,49 @@ function M.input(opts, on_confirm)
+---
+--- vim.ui.open_uri("/path/to/file")
+--- </pre>
+function M.open_uri(uri)
```
Perhaps we should just name it `vim.ui.open`, mentioning "URI" in the name is perhaps confusing because it also supported non-URI filesystem names. 

Opening stuff in the editor (Nvim) is called "edit" rather than "open" so that hopefully is enough distinction.

mrshmllow replied:
I think os_open is a good fit

justinmk commented on runtime/doc/news.txt:
```diff
@@ -45,6 +45,9 @@ iterators |luaref-in|.
 <
 • Enabled treesitter highlighting for treesitter query files by default.
 
+• Added |vim.ui.open_uri()| for opening URIs in the operating system's 
```
```suggestion
• |vim.ui.open_uri()| opens URIs using the system default handler (macOS `open`, Windows Explorer, Linux `xdg-open`, etc.) 
```

folke commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,49 @@ function M.input(opts, on_confirm)
+  elseif vim.fn.has("win32") then
+    cmd = "start"
+  else -- Assume linux
+    cmd = 'xdg-open'
```
There was also a good reason why I switched to `explorer` instead of `start` on Windows, but I can't remember why :)

mrshmllow replied:
I cannot find much of a reason to prefer explorer. `start` might even be better since it can provide better filetype-handler errors where as `explorer` will just fail silently and open in the file explorer in a random default location. (I think)

T-727 replied:
`start` is a "command" rather than an "executable". Using it instead of `explorer` here just errors out and does nothing

mrshmllow replied:
curious

T-727 replied:
If you're very keen on using `start` you could try to pass `cmd` + `path` as a joined string rather than a table I suppose. (`:help jobstart()` for full details)

folke commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,49 @@ function M.input(opts, on_confirm)
+    cmd = 'xdg-open'
+  end
+
+  local ret = vim.fn.system({
```
Better to use `jobstart` here. `vim.fn.system` sometimes fails for unknown reasons. See here for example https://github.com/folke/lazy.nvim/issues/18

```lua
  local ret = vim.fn.jobstart(cmd, { detach = true })
  if ret <= 0 then
    local msg = {
      "Failed to open uri",
      ret,
      vim.inspect(cmd),
    }
    vim.notify(table.concat(msg, "\n"), vim.log.levels.ERROR)
  end
```

folke replied:
I accidentally requestes changes, but these are obviously more suggestions

mrshmllow commented:
This would be a lot easier if i had a windows machine to test it on.

ii14 commented on runtime/lua/vim/ui.lua:
```diff
@@ -104,4 +104,57 @@ function M.input(opts, on_confirm)
+  end
+
+  local ret = vim.fn.jobstart(cmd, {
+    detatch = true,
```
```suggestion
    detach = true,
```

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2211,6 +2211,24 @@ input({opts}, {on_confirm})                                   *vim.ui.input()*
                       typed (it might be an empty string if nothing was
                       entered), or `nil` if the user aborted the dialog.
 
+os_open({path})                                             *vim.ui.os_open()*
```
After more thought, I'm pretty certain that `vim.ui.open` is good. There could be `vim.fs.open` also, but they're in separate namespaces and the purpose is clear: "fs" is for filesystem, "ui" is where platform-specific providers live.

justinmk replied:
```suggestion
open({path})                                             *vim.ui.open()*
```

justinmk committed:
fix(gx): visual selection, expand env vars

---
Rejected experiment: move vim.ui.open() to vim.env.open()

Problem:
`vim.ui` is where user-interface "providers" live, which can be
overridden. It would also be useful to have a "providers" namespace for
platform-specific features such as "open", clipboard, python, and the other
providers listed in `:help providers`. We could overload `vim.ui` to
serve that purpose as the single "providers" namespace, but
`vim.ui.nodejs()` for example seems awkward.

Solution:
`vim.env` currently has too narrow of a purpose. Overload it to also be
a namespace for `vim.env.open`.

diff --git a/runtime/lua/vim/_meta.lua b/runtime/lua/vim/_meta.lua
index 913f1fe20348..17d05ff37595 100644
--- a/runtime/lua/vim/_meta.lua
+++ b/runtime/lua/vim/_meta.lua
@@ -37,8 +37,28 @@ local options_info = setmetatable({}, {
   end,
 })

-vim.env = setmetatable({}, {
-  __index = function(_, k)
+vim.env = setmetatable({
+  open = setmetatable({}, {
+      __call = function(_, uri)
+        print('xxxxx'..uri)
+        return true
+      end,
+      __tostring = function()
+        local v = vim.fn.getenv('open')
+        if v == vim.NIL then
+          return nil
+        end
+        return v
+      end,
+    })
+  },
+  {
+  __index = function(t, k, ...)
+    if k == 'open' then
+      error()
+      -- vim.print({...})
+      -- return rawget(t, k)
+    end
     local v = vim.fn.getenv(k)
     if v == vim.NIL then
       return nil

justinmk commented:
pushed an update : 

- fix for visual `gx`
- test coverage
- buggy condition in lsp handler

justinmk committed:
fix(vim.ui.open): return (don't show) error message

Problem:
Showing an error via vim.notify() makes it awkward for callers such as
lsp/handlers.lua to avoid showing redundant errors.

Solution:
Return the message instead of showing it. Let the caller decide whether
and when to show the message.

mrshmllow commented:
I do hate `explorer` but `start` is somehow worse. currently with `explorer` on windows `vim.ui.open("google.com")` will open the file manager at some random location without warning.

But `start` requires a bunch of weird stuff. start for some reason interprets the first argument as the title, but not always(?). start also is entirely different depending on the shell and has some strangeness with spaces, which explorer properly handles.

justinmk commented:
> currently with `explorer` on windows `vim.ui.open("google.com")` will open the file manager at some random location without warning.

without "https://" that would be treated as a filepath on any system, afaik. 

If you find that netrw's `gx` on Windows works better, we could use its `start rundll32 url.dll,FileProtocolHandler` approach: https://github.com/neovim/neovim/blob/3ecd45ded044c47efa76b74e9e3b720fbe27adc7/runtime/autoload/netrw.vim#L5431-L5433

zeertzjq commented:
Works for me

zeertzjq commented:
Cannot reproduce

justinmk commented:
> the netrw plugin `share/nvim/runtime/plugin/netrwPlugin.vim` is being loaded before the `init.lua`file (where I tell nvim to not load netrw through the `vim.g.loaded_netrwPlugin` global var)

May need to revert https://github.com/neovim/neovim/pull/24262 , or move this mapping to https://github.com/neovim/neovim/blob/06c4edf46bf65ec6b540de355cf1a691533f1df9/runtime/lua/vim/_editor.lua#L1008

justinmk commented:
> I think that the reason is that Explorer doesn't accept the URL ...  Is there any alternative to avoid this dependency?

See above, https://github.com/neovim/neovim/pull/23401#issuecomment-1621187483

mrshmllow commented:
maybe it's worth switching the opener strategy on windows

justinmk commented:
is that different than what i said in https://github.com/neovim/neovim/pull/23401#issuecomment-1621187483 ?

mrshmllow commented:
#24392

adelarsq commented:
> is that different than what i said in https://github.com/neovim/neovim/pull/23401#issuecomment-1621187483 ?

It's the same issue.

> #24392

I had confirmed that with this fix `vim.ui.open` works as expected on Windows.

mrshmllow commented:
> May need to revert https://github.com/neovim/neovim/pull/24262 , or move this mapping to...

#24420
